Sync charts zoom in history tab

birrejan opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
Add synchronized zoom functionality across multiple charts in the History tab with a unified reset interface. When users zoom into one chart, all charts automatically synchronize to the same time range, providing a cohesive view of historical data across different entities and chart types.


https://github.com/user-attachments/assets/77510e16-87ed-40af-98dd-0e3138f4e4b5



## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #23766  
- This PR is related to issue or discussion:
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

birrejan committed:
Add chart zoom event system and sync infrastructure to chart-base

- Replace inline datazoom handler with dedicated _handleDataZoomEvent method
- Add _syncZoomState method for zoom state synchronization
- Refactor zoom detection to be more robust and reliable

birrejan committed:
Add hide reset button functionality

- Add hideResetButton property to ha-chart-base component
- Add hideResetButton property to state-history-chart-line component
- Add hideResetButton property to state-history-chart-timeline component
- Implement conditional reset button rendering based on hideResetButton flag
- Pass hideResetButton prop through component hierarchy

This allows parent components to control reset button visibility when
implementing custom reset functionality or coordinated multi-chart resets.

birrejan committed:
Implement chart zoom synchronization

- Add chart-zoom event handlers to state-history-chart-line component
- Add chart-zoom event handlers to state-history-chart-timeline component
- Forward zoom events with chart index for identification

This enables individual charts to communicate zoom changes to parent
components for coordinated multi-chart synchronization.

birrejan committed:
Add floating reset button and sync orchestration

- Add chart-zoom event type definition to HASSDomEvents interface
- Add global zoom state tracking with _hasZoomedCharts property
- Add _isSyncing flag to prevent infinite sync loops
- Implement _handleTimelineSync method for coordinating chart synchronization
- Implement _handleGlobalZoomReset method for resetting all charts
- Enable hide-reset-button on individual charts to use global reset
- Add floating reset button with Material Design styling

On history page the floating reset button appears when any chart is zoomed
and provides a single point to reset all synchronized charts simultaneously.

bramkragten commented on src/components/chart/state-history-charts.ts:
```diff
@@ -264,6 +296,85 @@ export class StateHistoryCharts extends LitElement {
+          return;
+        }
+
+        const chartBase =
+          chartComponent.renderRoot?.querySelector("ha-chart-base");
+
+        if (chartBase && chartBase.chart) {
+          chartBase.chart.dispatchAction({
```
Create a public zoom function on `ha-chart-base` and on `state-history-chart-line`, `state-history-chart-line` call the function on the latter, and let it forward it to `ha-chart-base`. `ha-chart-base` should do the `dispatchAction` call.

JanLayola replied:
that's indeed a cleaner solution.

birrejan committed:
Refactor chart zoom synchronization to use public API

Replace direct ECharts dispatchAction calls with proper zoom methods. The parent component now calls chartComponent.zoom() instead of accessing internal chart.dispatchAction() directly.

JanLayola commented on src/components/chart/state-history-charts.ts:
```diff
@@ -29,6 +31,13 @@ const chunkData = (inputArray: any[], chunks: number) =>
 declare global {
   interface HASSDomEvents {
     "y-width-changed": { value: number; chartIndex: number };
+    "chart-zoom": {
```
You're absolutely right, I missed the general overview. Now it's fix! Thanks

birrejan committed:
Remove duplicate TypeScript declaration of the "chart-zoom" event

birrejan committed:
Fix tooltips not shown due to xAxisPointer hidden

JanLayola commented:
@LiBe-net good catch! Now they should be working again

birrejan committed:
Use chart zoom function in history charts

bramkragten commented on src/components/chart/state-history-charts.ts:
```diff
@@ -115,19 +121,34 @@ export class StateHistoryCharts extends LitElement {
+            this._renderHistoryItem(item, index)
+          )}`}
+      ${this._hasZoomedCharts
+        ? html`<ha-icon-button
```
We should use a `ha-fab` here instead of an `icon-button`

bramkragten commented on src/components/chart/state-history-charts.ts:
```diff
@@ -264,6 +289,70 @@ export class StateHistoryCharts extends LitElement {
+    this._isSyncing = true;
+
+    requestAnimationFrame(() => {
+      const chartComponents = this.renderRoot.querySelectorAll(
+        "state-history-chart-line, state-history-chart-timeline"
+      );
```
```suggestion
      const chartComponents = this.renderRoot.querySelectorAll(
        "state-history-chart-line, state-history-chart-timeline"
      ) as StateHistoryChartLine | StateHistoryChartTimeline;
```

birrejan committed:
Apply code review feedback

birrejan committed:
Remove unnecessary any types

birrejan committed:
Apply code review feedback

MindFreeze commented on src/components/chart/state-history-chart-line.ts:
```diff
@@ -192,6 +197,21 @@ export class StateHistoryChartLine extends LitElement {
+      start: ev.detail.start ?? 0,
+      end: ev.detail.end ?? 100,
+      chartIndex: this.chartIndex,
+      startTime: ev.detail.startTime,
+      endTime: ev.detail.endTime,
```
startTime & endTime are not used anywhere

MindFreeze commented on src/components/chart/state-history-charts.ts:
```diff
@@ -29,6 +34,13 @@ const chunkData = (inputArray: any[], chunks: number) =>
+      start: number;
+      end: number;
+      chartIndex: number;
+      startTime?: Date;
+      endTime?: Date;
```
startTime & endTime are not used anywhere

MindFreeze commented on src/components/chart/state-history-charts.ts:
```diff
@@ -345,6 +442,12 @@ export class StateHistoryCharts extends LitElement {
     }
+    .reset-button {
+      position: fixed;
+      bottom: 24px;
+      right: 24px;
```
This is something we are adding right now so I don't expect you to know about it but may as well add it

```suggestion
      bottom: calc(24px + var(--safe-area-inset-bottom));
      right: calc(24px + var(--safe-area-inset-right));
```

MindFreeze reviewed (changes requested):
Found some issues:
- Double click to zoom out doesn't work. Probably a problem with the state of `this._isZoomed`
- Reset button is not rendered even when zoomed
- Ctrl/Cmd + drag with mouse to zoom doesn't sync and only zooms 1 chart

MindFreeze reviewed (approved):
Something must have been wrong with my build. Works great now. Thank you

MindFreeze commented:
@bramkragten do we still need UX input here ?

JanLayola commented:
I tested on a pixel 8 with all my entities (more than 500 hundred) and it's usable. Also I don't think it make sense to display that amount of data or more together but it works.

MindFreeze reviewed (approved):
Performance is about what you would expect in my testing as well. It's slow with a lot of charts but not because of the zoom sync. Confirmed by testing with the same entities on dev

MindFreeze commented:
Thanks you @JanLayola
